server/service/mdm.go

```diff
@@ -1371,10 +1392,39 @@ func (svc *Service) validateProfileLabels(ctx context.Context, labelNames []stri
 ////////////////////////////////////////////////////////////////////////////////
 
 type batchSetMDMProfilesRequest struct {
-	TeamID   *uint             `json:"-" query:"team_id,optional"`
-	TeamName *string           `json:"-" query:"team_name,optional"`
-	DryRun   bool              `json:"-" query:"dry_run,optional"` // if true, apply validation but do not save changes
-	Profiles map[string][]byte `json:"profiles"`
+	TeamID   *uint                        `json:"-" query:"team_id,optional"`
+	TeamName *string                      `json:"-" query:"team_name,optional"`
+	DryRun   bool                         `json:"-" query:"dry_run,optional"` // if true, apply validation but do not save changes
+	Profiles backwardsCompatProfilesParam `json:"profiles"`
+}
+
+type backwardsCompatProfilesParam []fleet.MDMProfileBatchPayload
+
+func (bcp *backwardsCompatProfilesParam) UnmarshalJSON(data []byte) error {
+	if len(data) == 0 {
+		return nil
+	}
+
+	// use []fleet.MDMProfileBatchPayload to prevent infinite recursion if
+	// we use `backwardsCompatProfileSlice`
+	var profs []fleet.MDMProfileBatchPayload
+	newFormatErr := json.Unmarshal(data, &profs)
+	if newFormatErr == nil {
+		*bcp = profs
+		return nil
+	}
+
+	var backwardsCompat map[string][]byte
+	oldFormatErr := json.Unmarshal(data, &backwardsCompat)
+	if newFormatErr != nil && oldFormatErr != nil {
+		return fmt.Errorf("unmarshal profile spec. Error using new format: %w. Error using old format: %w", newFormatErr, oldFormatErr)
+	}
+
+	*bcp = make(backwardsCompatProfilesParam, 0, len(backwardsCompat))
+	for name, contents := range backwardsCompat {
```

Comment:
Same here to avoid double-unmarshal we can look at first byte (if there's time, non critical).